### DISCUSSION TITLE
Champion "Const Var"

### DISCUSSION THREAD

**Discussion by gafter** [dotnet/csharplang#8663] [Language Ideas]
- [ ] Proposal added
- [ ] Discussed in LDM
- [ ] Decision in LDM
- [ ] Finalized (done, rejected, inactive)
- [ ] Spec'ed

See also https://github.com/dotnet/roslyn/issues/4423

**Comment by Pzixel** [comment]
most of use cases of `var` are anonymous types or very long names like `IGrouping<IEnumerable<IDictionary<...,....>, ...>>` and so on. There is no problem with writing `const int` or `const string` especially when people frequently has rules like `explicit declaration for primitive types`.  It's not a bad feature, but just useless, especially when you know that every feature starts with -100 points and you should outweight it with some benefit. I don't see any benefit here, until we can declare `const` of any type.

**Comment by CyrusNajmabadi** [comment]
> most of use cases of var are anonymous types or very long names like IGrouping<IEnumerable<IDictionary<...,....>, ...>> and so on

I don't believe there is any data that supports that statement.  Teams (Roslyn-IDE for one) may just choose to use 'var' for all types where it can be used.

**Comment by jnm2** [comment]
I use var for all locals.

**Comment by Pzixel** [comment]
@CyrusNajmabadi this feature is as simple as implement `_` to be a separator in numbers, but it just doesn't provide any expressiveness.

@jnm2 it's quite common to explicitly specify simple types. R# even has a rule for it
 
![image](https://cloud.githubusercontent.com/assets/11201122/24424710/470047ce-140a-11e7-8d5f-39f35760e43c.png)

I'm not saying it's completely useless, but here is bunch of much better champions.

**Comment by jnm2** [comment]
@Pzixel I'm aware of that, and I'm also aware that it's quite common not to specify simple types.

I bemoan the fact that `var` (signifying "variable") doesn't work well with `const`, just like it doesn't work well as a return type. But it's probably best to stick with `var` and redefine it mentally to mean "auto" rather than "variable." Just like "ReadOnly" now signifies "Readable" in .NET APIs. It's an artifact of history.

**Comment by weitzhandler** [comment]
`var` could be omitted.

**Comment by CyrusNajmabadi** [comment]
Working on PR here: https://github.com/dotnet/roslyn/pull/21149

**Comment by HekiShavik** [comment]
The `var` in `const var a = "A";` is moot in my oppinion. I find the EcmaScript `let a = "A";` and `const a = "A";` syntax quite natural - even with the very limited amout I write - and while I rarely use `const string a = "A";` in C#, I am completely on board with @weitzhandler and I wouldn't even think to use `const var a = "A";` unless Roslyn or ReSharper were extremely adamant about it. Even then I'd feel the `var` is very redundant and unnecessary to the point where I would refrain from using the `const` keyword at all in that connection simply because I feel it muddies my code.

I get why this isn't getting much attention because I only see the real benefit with enums:

`const TravellerBase.TravellerType typeFilter = TravellerBase.TravellerType.Adult | TravellerBase.TravellerType.Youth;`

versus

`const typeFilter = TravellerBase.TravellerType.Adult | TravellerBase.TravellerType.Youth;`

By the way, if [`using` gets to loose its brackets entirely](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/using), this would definitely be a piece of cake to implement and actually increase readability unlike the pattern based usings..

**Comment by hrumhurum** [comment]
Currently we can do `var s = "123"` but we cannot do `const s = "123"`.

Fields are in question, and generally speaking they should not be inferred for `const` as for now. Just to mirror the same choice made for `var`: vars are not allowed for fields.

`const var` syntax is out of question, as it represents a conflicting statement: the value cannot be a constant and a variable at the same time, it just does not make sense.

But local inferred const values are a way to go:

```csharp
void Main()
{
    const s = "Hello const!";
    Console.WriteLine(s);
}
```
That would be a great addition to C#. Personally I hit this omission several times a week.

**Comment by CyrusNajmabadi** [comment]
> const var syntax is out of question, 

It's definitely not out of the question :).  'var' doesn't mean 'the value varies'.  it means 'the type is inferred here'.  So `const var` means 'a constant, whose type is inferred'. 

`var` doesn't mean `it varies` anymore than `int i` means "i varies".  `var` is in place of the type, and only means `is inferred`.  That's why we don't say `var i: int`, just `var i`.

**Comment by Pzixel** [comment]
Interesting fact: Rust has `let` statement with optional type ascription for anything *but* constants, where it's required.

**Comment by CleanCodeX** [comment]
I find this apporach very useful. I don't care if it is const var or just const alone.

**Comment by hrumhurum** [comment]
`const var` needs further considerations.

An example:

```c#
int x = 10;
const var finalResult = x + 5;
```

Would this be a valid code? If not, what syntax would be used for immutable local variables if they ever appear in C#?

Would it be `sealed var`? Or `let`?

E.g.

```c#
int x = 10;
sealed var finalResult = x + 5;
```
vs
```c#
int x = 10;
let finalResult = x + 5;
```

If you are OK with `sealed var` long term then `const var` would be fine as well -- both constructions follow the same spirit.

Alternatively, if you prefer to choose a concise `let` keyword for immutable local variables then it should be just `const` for type-inferred constants to preserve the common language spirit.

Both approaches work. Here are some syntax samples for your convenience:

```c#
// ------------------------------------------
// 1. const var / sealed var approach
// ------------------------------------------
const var s = "Hello";
int x = 10;
sealed int y = x + 1; // type is specified
sealed var z = x + 2; // type is inferred

// ----------------------------------------
// 2. const / let approach
// ----------------------------------------
const s = "Hello";
int x = 10;
let int y = x + 1; // type is specified
let z = x + 2; // type is inferred
```

My personal vote goes for `const / let approach`, but `const var / sealed var` is not bad at all, despite being chatty and a bit confusing in "const var" statement from a human linguistics standpoint (just like a "black white" color).

**Comment by CleanCodeX** [comment]
My favorite ist just const with ommitting its type. const acts as var in that way.
const var seems to me to long for something const without type can do as well. I would not like to see sealed / let  to introduce new meanings or occurrences of these words.

**Comment by HaloFour** [comment]
@hrumhurum 

See https://github.com/dotnet/csharplang/issues/188

`const` means something very different to the C# compiler.  `const` locals don't exist in IL, and `const` fields aren't proper fields that can be referenced and read either.  The value must be evaluated at compile-time and embedded directly in any call or expression that uses it.

**Comment by weitzhandler** [comment]
My personal preference would be omitting the `var`. If you don't specifiy a type, you mean you want an inferred `const` type, though `var` can be allowed for those who like to see it.

And please don't introduce any new unrelated keywords like `let`, `sealed` or anything but `readonly` and `const` (both), where `readonly` is per scope and `const` is persistent/static.

**Comment by CleanCodeX** [comment]
> @hrumhurum
> 
> See #188
> 
> `const` means something very different to the C# compiler. `const` locals don't exist in IL, and `const` fields aren't proper fields that can be referenced and read either. The value must be evaluated at compile-time and embedded directly in any call or expression that uses it.

How the runtime treats the "const [type]"  expression and how it should "apear" to the developer in code is not necessarily the same. const without type could under the hood be compiled as const [type] as well.

**Comment by HaloFour** [comment]
@CleanCodeX 

C# already treats them differently, `readonly` vs `const`.  The latter has a _lot_ of restrictions since it must be evaluated by the compiler and it must be something that can be embedded in assembly metadata.  `const` expressions will never be evaluated at runtime, `readonly` expressions always will.

**Comment by CleanCodeX** [comment]
> @CleanCodeX
> 
> C# already treats them differently, `readonly` vs `const`. The latter has a _lot_ of restrictions since it must be evaluated by the compiler and it must be something that can be embedded in assembly metadata. `const` expressions will never be evaluated at runtime, `readonly` expressions always will.

I know. If const without type is limited to constant expressions, there is no reason not to ommit the type argument. the compiler can choose at design time which type it is. the things only get more difficult when the const value needs to be evaluated first.

**Comment by CleanCodeX** [comment]
I see no reason why:

const a = 3;
const b = "a";

cannot be compiled as 
const int a = 3;
const string b = "a";

if some wants a different type than the compiler chose, just type it. same applies for var, as well.

**Comment by weitzhandler** [comment]
> if some wants a different type than the compiler chose, just type it. same applies for var, as well.

Or if someone wants to 'declare' they're type-inferring they could choose to type out `const var = 3`, although that doesn't have to be mandatory.

**Comment by HaloFour** [comment]
@CleanCodeX 

I wasn't commenting on that specific syntax decision, only that `const` doesn't mean "immutable" and shouldn't be conflated with the other proposal to add such locals to the language.

**Comment by CleanCodeX** [comment]
> I wasn't commenting on that specific syntax decision, only that `const` doesn't mean "immutable" and shouldn't be conflated with the other proposal to add such locals to the language.

I aggree. Const for constant expressions, readonly for the others seems to address the mentioned dfficulties.

**Comment by hrumhurum** [comment]
Maybe 2023 will be finally the year of a const type inference in C#?

Writing long enums is especially annoying:

```c#
const DiscoveryOptions discoveryOptions = DiscoveryOptions.NoSort | DiscoveryOptions.Invariant;
```

instead of a more natural and concise statement below:

```c#
const discoveryOptions = DiscoveryOptions.NoSort | DiscoveryOptions.Invariant;
```

As it already had been mentioned, JavaScript has this feature since ES6 (2015) and it causes exactly zero problems. Why C# is still lagging on this front?

**Comment by theunrepentantgeek** [comment]
> JavaScript has this feature since ES6 (2015)

The keyword **const** has a very (very!) different meaning in JavaScript than it does in C#, plus the type systems are vastly different. 

> Why C# is still lagging on this front?

Every feature has costs and benefits - including the opportunity cost that working on one feature means a different feature doesn't get progressed. C# doesn't have this feature because there are other, more valuable features, that have been prioritised. 

That said, even in a mythical world of infinite resources, not every feature would make it into the language. The teams discipline in adding only the features that makes sense is to be applauded.

It's worth noting that this features has a champion, putting it ahead of many other features as someone on the C# team has decided this is valuable enough to progress further. Even that's no guarantee of progress though.

**Reply by fandrei** [reply]
@theunrepentantgeek 
> The teams discipline in adding only the features that makes sense is to be applauded.

Can't agree. Some of the features that _were_ implemented don't make any sense at all. So far, the new "from the end" array initialization by index is #1 in my anti-hit list.

**Comment by glen-84** [comment]
I have analyzers suggesting changing:

```cs
var key = "Foo";
```

To:

```cs
const string key = "Foo";
```

... but it's just so verbose.

I also think that, although `var` doesn't mean that the value varies, `const var` is still a bit awkward to read, and [could be shorter](https://github.com/dotnet/csharplang/issues/106#issuecomment-303028491).

```cs
const key = "Foo";
```

Are there any technical complications/ambiguities with this? Is this issue only blocked by available time and priorities?

**Comment by glen-84** [comment]
@colejohnson66 @Zodt What do you disagree with?

**Comment by TahirAhmadov** [comment]
I think we should `const x = 1;`, no need for `var` - I understand `var` means `auto`, but still - why type another keyword if it's not needed? I hope there are no parsing/lexical complications with expecting an identifier right after `const`...

**Comment by hrumhurum** [comment]
I just had a spontaneous revelation on this subject that came to me while I was working on a particular snippet of code. In general, it seems that there are too camps of thought:

1) `const var s = "abc"` syntax as proposed here
2) `const s = "abc"` syntax as proposed at https://github.com/dotnet/csharplang/issues/4242

The `const s = "abc"` syntax is purely a type inference for the existing constant definition syntax `const string s = "abc"`.

But the `const var s = "abc"` is something different. It is not clear what it tries to convey because a symbol cannot be a constant (const) and a variable (var) at the same time. Such thing cannot exist in the algebraic terms because `is_constant(symbol) != is_variable(symbol)` expression is always true in the existing language model. `const var s = "abc"` is an example of contradictory logic because it violates the axiom.

Now, to the matter of the question. In fact, C# misses two distinct features:

1) Type inference for the existing `const` keyword
2) Immutable variables

Those missing features can be naturally implemented using the following approaches:

1) Type inference for `const` keyword can be achieved using the natural `const s = "abc"` syntax which seems to be preferred by the majority of users:

      ```c#
      const PI = 3.1415926; // double
      const s = "abc"; // string
      const n = 123; // integer
    ```

2) Immutable variables can be naturally achieved by combining two existing keywords together producing `readonly <type> <name> = <value>` syntax. For example:

      ```c#
    void Method()
    {
          readonly DateTime now = DateTime.UtcNow;
    }
    ```
   As a direct consequence of that syntax, the following syntax becomes possible:
      ```c#
    void Method()
    {
          readonly var now = DateTime.UtcNow;
    }
    ```
   So now we have `readonly var` syntax which does not violate any algebraic preconditions of the language model.

By following those two synergetic and mathematically sound suggestions, the issue of missing language features will be finally solved.

**Reply by HaloFour** [reply]
See: [#8479](https://github.com/dotnet/csharplang/discussions/8479#discussioncomment-10837230)
